File: 15. 3Sum/Fixed/3sum.cpp
#include <iostream>
#define NOE 3

using namespace std;

int result_number = 0;
int number_of_elements = NOE;

class Result{
	int values [NOE];
	bool not_null;
	int number_of_values;
	
	public:
		Result(){
		}
		
		Result(int* list, int count){
			set_result (list, count);
			number_of_values = count;
			not_null = true;
		}
		
		void set_result(int* list, int count){
			for (int i = 0; i < count; i++){
				values[i] = list[i];
			}
		}
		
		void print_result (){
			cout << "This is the handy work of C++" << endl;
			for (int i = 0; i < number_of_elements; i++){
				cout << values[i];
				if (i < (number_of_values - 1)){
					cout << "\t";
				}
			}
			cout << endl;
		}
};

int factorial (int number){
	int out = 1;
	for (int i = number; i > 0; i--){
		out *= i;
	}
	
	return (out);
}

void sum_of_three (int* list, int len_list, int target, Result* out){
	int result_pointer = 0;
	
	for (int a = 0; a < len_list; a++){
		for (int b = a; b < len_list; b++){
			if (b > a){
				for (int c = b; c < len_list; c++){
					if (c > b){
						if ((a != b) && (a != c) && (b != c)){
							int aa = list[a];
							int bb = list[b];
							int cc = list[c];
							
							if ((aa != bb) && (aa != cc) && (bb != cc)){
								if ((aa + bb + cc) == target){
									int y[] = {aa, bb, cc};
									
									out[result_pointer] = Result(y, NOE);
									result_pointer++;
									result_number = result_pointer;
								}
							}
						}
					}
				}
			}
		}
	}
}

int main (){
	int nums[] = {-1,0,1,2,-1,-4};
	
	int len_nums = sizeof(nums)/sizeof(nums[0]);
	
	int sample_space = (factorial(len_nums) / (factorial(len_nums - number_of_elements) * factorial(number_of_elements)));
	
	Result h[sample_space];
	
	sum_of_three (nums, len_nums, 0, h);
	
	for (int i = 0; i < result_number; i++){
		h[i].print_result();
	}
	
	return (0);
}

File: 15. 3Sum/Fixed/3sum.rs
static mut result_number: isize = 0;
const number_of_elements: usize = 3;
const sample_size: usize = 10; 

#[derive(Clone, Copy)]
struct Result {
	values: [isize; number_of_elements],
	number_of_values: usize,
	not_null: bool
}

impl Result{
	
	fn new(list: &[isize]) -> Self{
		let mut arr: [isize; number_of_elements] = [0;number_of_elements];
		let count = list.len();
		
		for i in 0..count{
			arr[i as usize] = list[i as usize];
		}
		
		return Result{values: arr, number_of_values: count, not_null: true};
	}
	
	fn empty() -> Self{
		let arr: [isize; number_of_elements] = [0; number_of_elements];
		return Result{values: arr, number_of_values: number_of_elements, not_null: false};
	}
	
	fn set_result(&mut self, list: &[isize], count: usize) -> (){
		for i in 0..count{
			self.values[i as usize] = list[i as usize];
		}
	}
	
	fn print_result(self){
		println!("This is the handy work of Rust");
		for i in 0..self.number_of_values{
			print!("{}", self.values[i as usize]);
			if i < (self.number_of_values - 1){
				print!("\t");
			}
		}
		println!();
	}
}

fn factorial (number: isize) -> isize{
	let mut out: isize = 1;
	
	for i in (1..=number).rev(){
		out *= i;
	}
	
	return out;
}

fn sum_of_three (list: &[isize], target: isize, out: &mut[Result]) -> (){
	let mut result_pointer: isize = 0;
	let len_list: usize = list.len() as usize;
	
	for a in 0..len_list{
		for b in a..len_list{
			if b > a{
				for c in a..len_list{
					if c > b{
						if (a != b) && (a != c) && (b != c){
							let aa = list[a];
							let bb = list[b];
							let cc = list[c];
							
							if (aa != bb) && (aa != cc) && (bb != cc){
								if (aa + bb + cc) == target{
									out[result_pointer as usize] = Result::new(&[aa, bb, cc]);
									result_pointer += 1;
									unsafe {
										result_number = result_pointer;
									}
								}
							}
						}
					}
				}
			}
		}
	}
}

fn main() {
	//println!("{}\n\n\n", factorial(5));
	
	//let e: Result = Result{values: [2,3,4], number_of_values: 3, not_null: true};
	//let e: Result = Result::new(&[2,3,4]);
	//e.print_result();
	
	let nums: &[isize] = &[-1,0,1,2,-1,-4];
	
	let len_nums: usize = nums.len() as usize;
	
	let sample_space: isize = (factorial(len_nums as isize) / (factorial((len_nums - number_of_elements) as isize) * factorial(number_of_elements as isize)));
	
	if (sample_size < sample_space as usize){
		println!("The sample size of all combinations based of the size of the given array is less than the size allocated to handle the results of the program.\nKindly, within the program code file edit the sample size constant\n\"const sample_size: usize = {};\"\nto\n\"const sample_size: usize = {};\"\nPreferably a value that is greater than {}.\n", sample_size, sample_space, sample_space); 	
	} else {
		let def_Result: Result = Result::empty();
		
		let mut h: [Result; sample_size] = [def_Result; sample_size];
		
		sum_of_three(nums, 0, &mut h);
		
		let mut rn: isize;
		
		unsafe{
			rn = result_number;
		}
		
		for i in 0..rn{
			h[i as usize].print_result();
		}
	}
}
